lxd/instance/drivers/driver_lxc.go

```diff
@@ -4977,13 +4983,570 @@ func getCRIULogErrors(imagesDir string, method string) (string, error) {
+	d.logger.Debug("Waiting for migration completion response from target")
+	msg := migration.MigrationControl{}
+	err = args.ControlReceive(&msg)
+	if err != nil {
+		return fmt.Errorf("Failed receiving migration completion response from target: %w", err)
+	}
+
+	d.logger.Debug("Got migration completion response from target", logger.Ctx{"success": msg.GetSuccess(), "message": msg.GetMessage()})
+
+	if args.Live {
+		restoreSuccess <- msg.GetSuccess()
+		err := <-dumpSuccess
+		if err != nil {
+			d.logger.Error("Dump failed after successful restore", logger.Ctx{"err": err})
+		}
+	}
+
+	if !msg.GetSuccess() {
+		return fmt.Errorf(msg.GetMessage())
+	}
+
+	return nil
+}
+
+type preDumpLoopArgs struct {
+	liveConn      io.ReadWriteCloser
+	checkpointDir string
+	bwlimit       string
+	preDumpDir    string
+	dumpDir       string
+	final         bool
+	rsyncFeatures []string
+}
+
+// migrateSendPreDumpLoop is the main logic behind the pre-copy migration.
+// This function contains the actual pre-dump, the corresponding rsync transfer and it tells the outer loop to
+// abort if the threshold of memory pages transferred by pre-dumping has been reached.
+func (d *lxc) migrateSendPreDumpLoop(args *preDumpLoopArgs) (bool, error) {
+	// Do a CRIU pre-dump
+	criuMigrationArgs := instance.CriuMigrationArgs{
+		Cmd:          liblxc.MIGRATE_PRE_DUMP,
+		Stop:         false,
+		ActionScript: false,
+		PreDumpDir:   args.preDumpDir,
+		DumpDir:      args.dumpDir,
+		StateDir:     args.checkpointDir,
+		Function:     "migration",
+	}
+
+	d.logger.Debug("Doing another CRIU pre-dump", logger.Ctx{"preDumpDir": args.preDumpDir})
+
+	final := args.final
+
+	if d.Type() != instancetype.Container {
+		return false, fmt.Errorf("Instance is not container type")
+	}
+
+	err := d.migrate(&criuMigrationArgs)
+	if err != nil {
+		return final, fmt.Errorf("Failed sending instance: %w", err)
+	}
+
+	// Send the pre-dump.
+	ctName, _, _ := api.GetParentAndSnapshotName(d.Name())
+	err = rsync.Send(ctName, shared.AddSlash(args.checkpointDir), args.liveConn, nil, args.rsyncFeatures, args.bwlimit, d.state.OS.ExecPath)
+	if err != nil {
+		return final, err
+	}
+
+	// The function readCriuStatsDump() reads the CRIU 'stats-dump' file
+	// in path and returns the pages_written, pages_skipped_parent, error.
+	readCriuStatsDump := func(path string) (uint64, uint64, error) {
+		// Get dump statistics with crit
+		dumpStats, err := crit.GetDumpStats(path)
+		if err != nil {
+			return 0, 0, fmt.Errorf("Failed to parse CRIU's 'stats-dump' file: %w", err)
+		}
+
+		return dumpStats.GetPagesWritten(), dumpStats.GetPagesSkippedParent(), nil
+	}
+
+	// Read the CRIU's 'stats-dump' file
+	dumpPath := shared.AddSlash(args.checkpointDir)
+	dumpPath += shared.AddSlash(args.dumpDir)
+	written, skippedParent, err := readCriuStatsDump(dumpPath)
+	if err != nil {
+		return final, err
+	}
+
+	totalPages := written + skippedParent
+	percentageSkipped := int(100 - ((100 * written) / totalPages))
```

Comment:
Can we be sure that `totalPages` is always `> 0`?

Reply:
Hrm, I'm not sure (as this is old code), but I'll add a conditional around it to be safe, thanks.